src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/RestoreCommandHandler.cs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.ServiceHub.Framework;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Shell.ServiceBroker;
+using Microsoft.VisualStudio.Threading;
+using NuGet.VisualStudio.Internal.Contracts;
+using IAsyncServiceProvider = Microsoft.VisualStudio.Shell.IAsyncServiceProvider;
+using Task = System.Threading.Tasks.Task;
+
+namespace NuGet.VisualStudio.OnlineEnvironment.Client
+{
+    internal class RestoreCommandHandler
+    {
+        private readonly JoinableTaskFactory _joinableTaskFactory;
+        private readonly IAsyncServiceProvider _asyncServiceProvider;
+        private CancellationTokenSource _cancelBuildToken;
+
+        public RestoreCommandHandler(JoinableTaskFactory joinableTaskFactory, IAsyncServiceProvider asyncServiceProvider)
+        {
+            _joinableTaskFactory = joinableTaskFactory ?? throw new ArgumentNullException(nameof(joinableTaskFactory));
+            _asyncServiceProvider = asyncServiceProvider ?? throw new ArgumentNullException(nameof(asyncServiceProvider));
+        }
+
+        public void RunSolutionRestore()
+        {
+            _joinableTaskFactory.RunAsync(RunSolutionRestoreAsync);
+        }
+
+        private async Task RunSolutionRestoreAsync()
+        {
+            var cancellationTokenSource = new CancellationTokenSource();
+            _cancelBuildToken = cancellationTokenSource;
+
+            try
+            {
+                await GetAndActivatePackageManagerOutputWindowAsync();
+
+                await TaskScheduler.Default;
+                IBrokeredServiceContainer serviceContainer = await _asyncServiceProvider.GetServiceAsync<SVsBrokeredServiceContainer, IBrokeredServiceContainer>();
+                IServiceBroker serviceBroker = serviceContainer.GetFullAccessServiceBroker();
+
+                INuGetSolutionService nugetSolutionService = await serviceBroker.GetProxyAsync<INuGetSolutionService>(NuGetServices.NuGetSolutionService);
+
+                try
+                {
+                    await nugetSolutionService.RestoreSolutionAsync(cancellationTokenSource.Token);
+                }
+                finally
+                {
+                    (nugetSolutionService as IDisposable)?.Dispose();
+                }
+            }
+            catch (Exception e)
+            {
+                // Only log to the activity log for now
+                // TODO: https://github.com/NuGet/Home/issues/9352
+                ActivityLog.LogError("NuGet Package Manager", e.Message);
+            }
+            finally
+            {
+                cancellationTokenSource.Dispose();
```

Comment:
I'm trying to remember, is it necessary to set cancellation on the token source before disposing, to make sure that anything waiting on the cancellation token doesn't deadlock?

Reply:
I've not encountered that before. 
You want to make sure you release all your references so it gets gced.